core_lib/src/interface/backupelement.h

```diff
@@ -19,75 +20,162 @@ GNU General Public License for more details.
 #define BACKUPELEMENT_H
 
 #include <QObject>
-#include "vectorimage.h"
-#include "bitmapimage.h"
-#include "soundclip.h"
+#include <QUndoCommand>
+#include <QRectF>
+#include <QTransform>
 
-class Editor;
+#include "movemode.h"
+#include "pencildef.h"
+#include "layer.h"
+#include "vectorselection.h"
+#include "preferencemanager.h"
 
-class BackupElement : public QObject
+class Editor;
+class BackupManager;
+class PreferenceManager;
+class BitmapImage;
+class VectorImage;
+class SoundClip;
+class Camera;
+class Layer;
+class KeyFrame;
+class TransformElement;
+
+enum types { UNDEFINED,
+             ADD_KEY_MODIF,
+             REMOVE_KEY_MODIF
+           };
+
+class BackupElement : public QUndoCommand
 {
-    Q_OBJECT
 public:
-    enum types { UNDEFINED, BITMAP_MODIF, VECTOR_MODIF, SOUND_MODIF };
-
-    QString undoText;
-    bool somethingSelected = false;
-    qreal rotationAngle = 0.0;
-    qreal scaleX = 1.0;
-    qreal scaleY = 1.0;
-    QPointF translation;
-    QRectF mySelection;
-    QPointF selectionAnchor;
-
-    virtual int type() { return UNDEFINED; }
-    virtual void restore(Editor*) { Q_ASSERT(false); }
+    explicit BackupElement(Editor* editor, QUndoCommand* parent = nullptr);
+    virtual ~BackupElement();
+
+    Editor* editor() { return mEditor; }
+
+    bool isFirstRedo() const { return mIsFirstRedo; }
+    void setFirstRedo(bool state) { mIsFirstRedo = state; }
+
+    virtual int type() const { return UNDEFINED; }
+    virtual void undo() { Q_ASSUME(true); } // should never end here
+    virtual void redo() { Q_ASSUME(true); } // should never end here
+private:
+    Editor* mEditor = nullptr;
+    bool mIsFirstRedo = true;
 };
 
-class BackupBitmapElement : public BackupElement
+class BitmapElement : public BackupElement
 {
-    Q_OBJECT
+
 public:
-    explicit BackupBitmapElement(BitmapImage* bi) { bitmapImage = *bi; }
+    BitmapElement(const BitmapImage* backupBitmap,
+                  const int backupLayerId,
+                  Editor* editor,
+                  QString description,
+                  QUndoCommand* parent = nullptr);
 
-    int layerId = 0;
+    void undo() override;
+    void redo() override;
 
-    int layer = 0;
-    int frame = 0;
-    BitmapImage bitmapImage;
-    int type() override { return BackupElement::BITMAP_MODIF; }
-    void restore(Editor*) override;
+private:
+    int oldFrameIndex = 0;
+    int newFrameIndex = 0;
+
+    int oldLayerId = 0;
+    int newLayerId = 0;
+
+    BitmapImage* oldBitmap = nullptr;
+    BitmapImage* newBitmap = nullptr;
```

Comment:
Could you please give some more details about the differences between the new and old bitmap?

Reply:
I recently renamed the variables to old -> undo and new -> redo, hopefully that makes it more clear.